CONTRIBUTING.md

```diff
@@ -1,136 +1,230 @@
+https://github.com/Qiskit/qiskit-terra/pulls
+to review any open pull requests and provide feedback on it.
+
+### Good first contributions
+
+If you would like to contribute to the qiskit-terra but aren't sure of where
+to get started, the
+[`good first issue`](
+https://github.com/Qiskit/qiskit-terra/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)
+label in the issues is a good place to start. These are all issues that have
+been reviewed by contributors and tagged as something a new contributor
+should be able to develop a fix for. In other words it shouldn't require
+intimate familiarity with qiskit-terra to develop a fix for the issue.
+
+### Documentation
+
+If you make a change, make sure you update the associated
+*docstrings* and parts of the
+[documentation](https://github.com/Qiskit/qiskit/tree/master/docs/terra)
+that corresponds to it.
+
+### Pull requests
+
+We use [GitHub pull requests](
+https://help.github.com/articles/about-pull-requests) to accept the contributions.
+
+While not required, it is often good practice to open a new issue about
+the bug you're fixing or the feature you're implementing prior to creating
+a pull request. This will enable a discussion about what is being implemented
+and how best we can work together to implement it in the code.
 
-If you have an idea for a new feature please open a ticket labeled as
-`enhancement`. If you could also add a piece of code with the idea or a
-partial implementation it would be awesome.
+If you've written some code but need help finishing it, want to get initial
+feedback on it priort to finishing it, or want to share it and discuss prior
+to finishing the implementation you can open a *Work in Progress* pull request.
+When you create the pull request prefix the title with the **\[WIP\]** tag (for
+work in progress). This will indicate to reviewers that the code in the PR
+isn't in it's final state and will change. It also means that we will not merge
+the commit until it is finished. You or a reviewer can remove the [WIP] tag when
+the code is ready to fully reviewed for merging.
 
 ### Contributor License Agreement
 
-We\'d love to accept your code! Before we can, we have to get a few
-legal requirements sorted out. By signing a contributor license
-agreement (CLA), we ensure that the community is free to use your
-contributions.
+Before you can submit any code we need all contributors to sign a
+contributor license agreement. By signing a contributor license
+agreement (CLA) you're basically just an attesting to the fact
+that you are the author of the contribution and that you're freely
+contributing it under the terms of the Apache-2.0 license.
 
 When you contribute to the Qiskit Terra project with a new pull request,
 a bot will evaluate whether you have signed the CLA. If required, the
 bot will comment on the pull request, including a link to accept the
 agreement. The [individual CLA] document is available for review as a PDF.
```

Comment:
It shouldn't be needed because the links are included in the bottom. But I prefer the inline links since it is more clear in the raw format where it's being linked to. The only time we should use the bottom links is probably if there are multiple links to the same url.